include/sys/dnode.h

```diff
@@ -268,6 +269,29 @@ typedef struct dnode_phys {
 	};
 } dnode_phys_t;
 
+/*
+ * This structure holds only the portable fields of a dnode_phys_t, i.e.
+ * all fields but excluding the variable sized tail region. It is used to
+ * calculate the HMAC of a dnode_phys_t in zio_crypt_do_dnode_hmac_updates().
```

Comment:
It seems a little unfortunate to have to duplicate this.  Could we embed it in the dnode_phys_t instead?  Though I guess since it's on-disk it won't be changing very often.

Reply:
Yes, that was my concern and reasoning as well. I do type embedding frequently in Go but I failed to figure how to do it in C without requiring  Microsoft extensions (gcc/clang `-fms-extensions` flag). Even in C11 you can't have an anonymous struct
member of a previously defined type. Am I missing something obvious here? Or are you thinking of doing some `#define` ?

Reply:
I think the cleanest way is to do something like
```
struct dnode_phys {
    dnode_phys_core_t dn_core;
    union {
        ...
   }
}
```

The downside is that you have to change all the code that's accessing it to add `.dn_core` :( .  (And the "solution" to that of doing `#define dn_type dn_core.dn_type` seems worse.)

I'm guessing that the compiler's complaint about the current code is that you have a `dnode_phys_t*` that points to an array that is not large enough to hold a dnode_phys_t.  I think the proposed solution is decent but I wonder if there's another way to do this.  Maybe if we allocated a larger buffer to point to (might have to be on the heap due to the linux kernel's limited stack space).  That would be a little slower but it might not be significantly so.

Reply:
> The downside is that you have to change all the code that's accessing it to add `.dn_core` :( . (And the "solution" to that of doing `#define dn_type dn_core.dn_type` seems worse.)

Yeah, fully agree. It would be so nice if we could just say
```
struct dnode_phys {
    struct dnode_phys_core;
    union {
        ...
   }
}
```

but well, maybe in ten years, using C23, that'll work. ;-)

> I'm guessing that the compiler's complaint about the current code is that you have a `dnode_phys_t*` that points to an array that is not large enough to hold a dnode_phys_t.

Exactly, `‘dnode_phys_t {aka struct dnode_phys}[0]’ is partly outside array bounds`. The full warning is [here](https://github.com/openzfs/zfs/pull/12237#issuecomment-860797080) at the bottom.

>  Maybe if we allocated a larger buffer to point to

That didn't occur to me, but your right, using a `dnode_phys_t` and just limiting the HMAC calculation to the first 64 bytes of that may be a cleaner solution. IIRC the stack size in Linux is 16k and a `dnode_phys_t` is 512 bytes,  so limited stack space shouldn't be a concern. The only drawbacks I can see is 448 more bytes on the stack and maybe slightly less clearer code. I'll push a changed version. Which approach to choose I leave up to you, I'm fine either way.